src/refdb_fs.c

```diff
@@ -1416,26 +1439,14 @@ static int refdb_fs_backend__delete_tail(
 	else if (error == 0)
 		loose_deleted = 1;
 
-	if ((error = packed_reload(backend)) < 0)
-		goto cleanup;
-
-	/* If a packed reference exists, remove it from the packfile and repack */
-	if ((error = git_sortedcache_wlock(backend->refcache)) < 0)
+	if ((error = packed_delete(backend, ref_name)) < 0 && error != GIT_ENOTFOUND)
 		goto cleanup;
 
-	if (!(error = git_sortedcache_lookup_index(
-			&pack_pos, backend->refcache, ref_name)))
-		error = git_sortedcache_remove(backend->refcache, pack_pos);
-
-	git_sortedcache_wunlock(backend->refcache);
-
 	if (error == GIT_ENOTFOUND) {
 		error = loose_deleted ? 0 : ref_error_notfound(ref_name);
 		goto cleanup;
 	}
```

Comment:
I think this would become clearer if it was rolled into the error handling of `packed_delete`:

```
if ((error = packed_delete(backend, ref_name)) < 0) {
    if (error == GIT_ENOTFOUND)
       error = loose_deleted ? 0 : ref_error_notfound(ref_name);
    goto cleanup;
}
```

Not required, so please only consider this if you're re-rolling anyway

Reply:
Re-reading, I can't actually do that, as only `_delete_tail` has kept around whether a loose ref was existing at the start of the deletion. It turns out, #5003 clarifies what the correct ordering of operation is, so I've updated the PR to fix it (though I have a feeling we're still not quite there, see the missing (?) `packed-ref.lock` handling).